File: crates/viben-core/src/channels/slack.rs
//! Slack Channel Client
//!
//! Sends messages via Slack Web API

use super::types::{SendMessageOptions, SendMessageResult, SlackConfig, TestChannelResult};
use reqwest::Client;
use serde::{Deserialize, Serialize};

const SLACK_API_BASE: &str = "https://slack.com/api";

#[derive(Debug, Serialize)]
struct SendMessageRequest {
    channel: String,
    text: String,
}

#[derive(Debug, Deserialize)]
struct SlackResponse {
    ok: bool,
    error: Option<String>,
    ts: Option<String>,
}

#[derive(Debug, Deserialize)]
struct AuthTestResponse {
    ok: bool,
    error: Option<String>,
    user: Option<String>,
    team: Option<String>,
}

/// Send a message to a Slack channel
pub async fn send_slack_message(
    config: &SlackConfig,
    options: &SendMessageOptions,
) -> SendMessageResult {
    let token = match &config.token {
        Some(t) if !t.is_empty() => t,
        _ => return SendMessageResult::error("Bot token is required"),
    };

    if options.chat_id.is_empty() {
        return SendMessageResult::error("Channel ID is required");
    }

    let client = Client::new();

    let request_body = SendMessageRequest {
        channel: options.chat_id.clone(),
        text: options.message.clone(),
    };

    let response = match client
        .post(format!("{}/chat.postMessage", SLACK_API_BASE))
        .header("Authorization", format!("Bearer {}", token))
        .header("Content-Type", "application/json")
        .json(&request_body)
        .send()
        .await
    {
        Ok(r) => r,
        Err(e) => return SendMessageResult::error(e.to_string()),
    };

    let data: SlackResponse = match response.json().await {
        Ok(d) => d,
        Err(e) => return SendMessageResult::error(format!("Failed to parse response: {}", e)),
    };

    if !data.ok {
        return SendMessageResult::error(
            data.error.unwrap_or_else(|| "Unknown error".to_string())
        );
    }

    SendMessageResult::success(data.ts)
}

/// Test Slack channel configuration by calling auth.test
pub async fn test_slack_channel(config: &SlackConfig) -> TestChannelResult {
    let token = match &config.token {
        Some(t) if !t.is_empty() => t,
        _ => return TestChannelResult::error("Bot token is required"),
    };

    let client = Client::new();

    let response = match client
        .post(format!("{}/auth.test", SLACK_API_BASE))
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await
    {
        Ok(r) => r,
        Err(e) => return TestChannelResult::error(e.to_string()),
    };

    let data: AuthTestResponse = match response.json().await {
        Ok(d) => d,
        Err(e) => return TestChannelResult::error(format!("Failed to parse response: {}", e)),
    };

    if !data.ok {
        return TestChannelResult::error(
            data.error.unwrap_or_else(|| "Unknown error".to_string())
        );
    }

    TestChannelResult::success(format!(
        "Bot: {} (Team: {})",
        data.user.unwrap_or_else(|| "unknown".to_string()),
        data.team.unwrap_or_else(|| "unknown".to_string())
    ))
}

File: packages/core/src/channels/slack.ts
/**
 * Slack channel implementation
 *
 * Supports sending messages via Slack Web API.
 */

import type { SlackChannelConfig, SendMessageOptions, SendMessageResult, TestChannelResult } from "./types";
import { fetchWithProxy } from "./http-client";

/**
 * Slack API response types
 */
interface SlackAuthTestResponse {
  ok: boolean;
  error?: string;
  user?: string;
  team?: string;
  team_id?: string;
}

interface SlackPostMessageResponse {
  ok: boolean;
  error?: string;
  ts?: string;
  channel?: string;
}

/**
 * Send a message via Slack Web API
 */
export async function sendSlackMessage(
  config: SlackChannelConfig,
  options: SendMessageOptions
): Promise<SendMessageResult> {
  if (!config.token) {
    return { success: false, error: "Bot token is required" };
  }

  try {
    const response = await fetchWithProxy("https://slack.com/api/chat.postMessage", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${config.token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        channel: options.chatId,
        text: options.message,
        // Support markdown formatting
        mrkdwn: options.parseMode === "markdown",
      }),
    });

    const data = (await response.json()) as SlackPostMessageResponse;

    if (data.ok) {
      return {
        success: true,
        messageId: data.ts,
      };
    }

    return {
      success: false,
      error: data.error || "Unknown error",
    };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Test Slack channel configuration
 */
export async function testSlackChannel(
  config: SlackChannelConfig
): Promise<TestChannelResult> {
  if (!config.token) {
    return { success: false, error: "Bot token is required" };
  }

  try {
    const response = await fetchWithProxy("https://slack.com/api/auth.test", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${config.token}`,
        "Content-Type": "application/json",
      },
    });

    const data = (await response.json()) as SlackAuthTestResponse;

    if (data.ok) {
      const details = [
        data.user && `User: ${data.user}`,
        data.team && `Team: ${data.team}`,
      ]
        .filter(Boolean)
        .join(", ");

      return {
        success: true,
        details: details || "Connection successful",
      };
    }

    return {
      success: false,
      error: data.error || "Unknown error",
    };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Get Slack channel info
 */
export async function getSlackChannelInfo(
  config: SlackChannelConfig,
  channelId: string
): Promise<{ name?: string; error?: string }> {
  if (!config.token) {
    return { error: "Bot token is required" };
  }

  try {
    const response = await fetchWithProxy(
      `https://slack.com/api/conversations.info?channel=${encodeURIComponent(channelId)}`,
      {
        headers: {
          Authorization: `Bearer ${config.token}`,
        },
      }
    );

    const data = (await response.json()) as {
      ok: boolean;
      error?: string;
      channel?: { name: string };
    };

    if (data.ok && data.channel) {
      return { name: data.channel.name };
    }

    return { error: data.error || "Unknown error" };
  } catch (error) {
    return {
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * List Slack channels
 */
export async function listSlackChannels(
  config: SlackChannelConfig
): Promise<{ channels?: Array<{ id: string; name: string }>; error?: string }> {
  if (!config.token) {
    return { error: "Bot token is required" };
  }

  try {
    const response = await fetchWithProxy(
      "https://slack.com/api/conversations.list?types=public_channel,private_channel",
      {
        headers: {
          Authorization: `Bearer ${config.token}`,
        },
      }
    );

    const data = (await response.json()) as {
      ok: boolean;
      error?: string;
      channels?: Array<{ id: string; name: string }>;
    };

    if (data.ok && data.channels) {
      return {
        channels: data.channels.map((c) => ({
          id: c.id,
          name: c.name,
        })),
      };
    }

    return { error: data.error || "Unknown error" };
  } catch (error) {
    return {
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}
